### ISSUE TITLE
NetCoreApp2.0 support?

### ISSUE THREAD

**Issue by thisispaulsmith** [structuremap/structuremap#603]
Is NetCoreApp2.0 supported?

Reason for question is that I've noticed Registries aren't getting populated after shifting to 2.0.

Sample repo
https://github.com/thisispaulsmith/StructureMapNetCore2.0

Switching the test project to 1.1, the tests then work.

**Comment by jeremydmiller** [comment]
I haven't touched netcoreapp2.0 yet. No clue why that would even matter since SM is just a library that already targets netstandard. From your code, it's easily possible that SM's logic for walking up the call stack to find the calling assembly broke in 2.0. That changed quite a bit earlier. Try explicitly specifying the assembly in your Scan() method.

**Comment by thisispaulsmith** [comment]
Yes it would appear explicitly defining the assembly works

**Comment by silkfire** [comment]
Having the same issue. `TheCallingAssembly()` method in netcoreapp2.0 mysteriously returns the StructureMap assembly instead of my MVC project assembly.

And for some reason it says assembly version 4.5.0.0 instead of 4.5.2.0 (latest version of SM).

http://imgur.com/a/RgtEg

**Comment by jeremydmiller** [comment]
Guys, it's gonna be like this. There's an easy workaround, or I'll happily wait for a PR from somebody on this one.

**Comment by silkfire** [comment]
I checked the code [here](https://github.com/structuremap/structuremap/blob/master/src/StructureMap/Graph/AssemblyScanner.cs#L326) and right now the way `TheCallingAssembly()` determines the calling assembly based on the stack trace seems a bit arbitrary. You seem to assume that the calling assembly will always end up on the fifth row, but this case is no longer true in netcoreapp2.0.

Perhaps another approach should be devised in order to determine the calling assembly with a higher accuracy?

In case you wonder what has changed, `Environment.StackTrace` in turn calls `Environment.GetStackTrace()` prior to netcoreapp2.0; this is no longer the case - the logic has most probably been incorporated directly into the auto property. Thus `parts[4]` has become `parts[3]` in the current assembly scanner logic.

**Comment by volak** [comment]
Yeah in 2.0 there are a few things to be aware of when trying to assembly scanning.
First of all `TheCallingAssembly` doesn't work - just forget about it.
Second of all building doesn't put all nuget dependencies into the build directory - so having structuremap scan the directory won't work either. (only `dotnet publish` puts all the nuget depends into a release folder)

Workaround:

Put `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>` into your project settings - which will have the build move all the nuget packages over which in turn will make the directory scanning work.

The option will slow the build down and is not ideal.  I had a discussion on [gitter about this](https://gitter.im/dotnet/coreclr?at=5a26f862c65707ba2b16e743) and we found an internal process `AppContext.GetData("TRUSTED_PLATFORM_ASSEMBLIES")` which will return a list of assemblies in the app's *.deps.json - the file thats build during debugging.
In an ideal work structuremap could use that when doing "AssembliesFromBaseDirectory" so the project switch is not needed and scanning works when debugging.

This is what the folks at NServiceBus do https://github.com/Particular/NServiceBus/pull/5081

I might be up for taking a stab at incorporating this - but I'm pretty swamped so if its harder than a few hours work I might not get it done.

**Comment by volak** [comment]
Well - adding the code is not the hard part, but I'd have to upgrade the project to add a netstandard 2.0 target which I can't seem to get working right and don't know if that's acceptable or not.
I'm not aware of a way to get assemblies from the deps.json file in anything < standard 2.0 - if someone else knows let me know and I can do a PR
